pipeline_dp/dp_computations.py

```diff
@@ -123,16 +129,59 @@ def _add_random_noise(value: float, eps: float, delta: float,
         The value resulted after adding the random noise.
     """
     if noise_kind == pipeline_dp.NoiseKind.LAPLACE:
-        l1_sensitivity = compute_l1_sensitivity(l0_sensitivity,
-                                                linf_sensitivity)
+        l1_sensitivity = compute_l1_sensitivity(l0_sensitivity, linf_sensitivity)
         return apply_laplace_mechanism(value, eps, l1_sensitivity)
     if noise_kind == pipeline_dp.NoiseKind.GAUSSIAN:
-        l2_sensitivity = compute_l2_sensitivity(l0_sensitivity,
-                                                linf_sensitivity)
+        l2_sensitivity = compute_l2_sensitivity(l0_sensitivity, linf_sensitivity)
         return apply_gaussian_mechanism(value, eps, delta, l2_sensitivity)
     raise ValueError("Noise kind must be either Laplace or Gaussian.")
 
 
+@dataclass
+class AdditiveVectorNoiseParams:
+    eps: float
+    delta: float
+    max_norm: float
+    l0_sensitivity: float
+    linf_sensitivit: float
+    norm_kind: str
+    noise_kind: pipeline_dp.NoiseKind
+
+
+def _clip_vector(vec_sum: np.ndarray, max_norm: float, norm_kind="linf"):
+    if norm_kind == "linf":
+        return np.clip(vec_sum, -max_norm, max_norm)
+    elif norm_kind in ["l1", "l2"]:
+        norm_kind = int(norm_kind[-1])
+        vec_norm = np.linalg.norm(vec_sum, ord=norm_kind)
+        mul_coef = min(1, max_norm / vec_norm)
+        return vec_sum * mul_coef
+
+
+def add_noise_vector(vec_sum: np.ndarray, noise_params: AdditiveVectorNoiseParams):
+    """Adds noise to vector to make the vector sum computation (eps, delta)-DP.
+
+    Args:
+        vec_sum: the queried raw vector sum
+        noise_params: parameters of the noise to add to the computation
+    """
+    vec_sum = _clip_vector(vec_sum, noise_params.max_norm, noise_params.norm_kind)
+    vec_sum = np.array(
+        [
+            _add_random_noise(
+                s,
+                noise_params.eps,
+                noise_params.delta,
+                noise_params.l0_sensitivity,
+                noise_params.linf_sensitivit,
+                noise_params.noise_kind,
+            )
+            for s in vec_sum
+        ]
+    )
```

Comment:
@dvadym 
Shouldn't we request budget for each of the columns/features of the vector separately? If I understand correctly, each sum of a column is an independent query of the data and so having (e, d)-DP on each of the columns means (n*e, n*d)-DP for the overall vector sum (where n is the number of columns). Am I missing something?

Reply:
Yes, each column is an independent query. Let us assume that eps/delta in `AdditiveVectorNoiseParams` are per column (please rename to something like `eps_per_coordinate`, `delta_per_coordinate`).

Explanation: spliting of budget can be done in more effective way than just dividing by number of columns. This is pretty complicated and it's up to `BudgetAccountant` to do this.